src/features/world/scenes/BaseScene.ts

```diff
@@ -9,24 +9,98 @@ import { INITIAL_BUMPKIN } from "features/game/lib/constants";
 import { BumpkinContainer } from "../containers/BumpkinContainer";
 import { interactableModalManager } from "../ui/InteractableModals";
 import {
+  ChatContext,
   ChatMessageReceived,
   MachineInterpreter,
   PlayerJoined,
   PlayerQuit,
+  RoomEvent,
+  RoomId,
 } from "../roomMachine";
 import { CONFIG } from "lib/config";
 import { NPCName, NPC_WEARABLES } from "lib/npcs";
 import { npcModalManager } from "../ui/NPCModals";
 import { BumpkinParts } from "lib/utils/tokenUriBuilder";
 import { SUNNYSIDE } from "assets/sunnyside";
+import { EventObject, State } from "xstate";
 
 export type NPCBumpkin = {
   x: number;
   y: number;
   npc: NPCName;
 };
 
-export class BaseScene extends Phaser.Scene {
+export abstract class BaseScene extends Phaser.Scene {
+  abstract roomId: RoomId;
+  eventListener: (event: EventObject) => void;
+  transitionListener: (
+    state: State<ChatContext, RoomEvent, any, any, any>,
+    event: RoomEvent
+  ) => void;
+
+  constructor(key: string) {
+    super(key);
+
+    this.eventListener = (event) => {
+      if (event.type === "CHAT_MESSAGE_RECEIVED") {
+        const { sessionId, text } = event as ChatMessageReceived;
+        if (
+          sessionId &&
+          String(sessionId).length > 4 &&
+          this.playerEntities[sessionId]
+        ) {
+          this.playerEntities[sessionId].speak(text);
+        }
+      }
+
+      if (event.type === "PLAYER_JOINED") {
+        const { sessionId, x, y, clothing, roomId } = event as PlayerJoined;
+
+        if (roomId !== this.roomId) return;
+
+        const room = this.roomService.state.context.rooms[roomId];
+
+        if (!room) return;
+
+        console.log("Player joined", sessionId, x, y, clothing);
+        const player = this.createPlayer({
+          x,
+          y,
+          clothing,
+          isCurrentPlayer: sessionId === room.sessionId,
+        });
+        this.playerEntities[sessionId] = player;
+      }
+
+      if (event.type === "PLAYER_QUIT") {
+        const { sessionId, roomId } = event as PlayerQuit;
+
+        if (roomId !== this.roomId) return;
+
+        const entity = this.playerEntities[sessionId];
+        if (entity) {
+          entity.destroy();
+          delete this.playerEntities[sessionId];
+        }
+      }
+    };
+
+    this.transitionListener = (state) => {
+      if (state.value === "error" && !this.readonly) {
+        console.log("We have an error");
+        // Render the player for readonly
+        this.createPlayer({
+          x: 300,
+          y: 300,
+          isCurrentPlayer: true,
+          clothing: INITIAL_BUMPKIN.equipped,
+        });
+
+        this.readonly = true;
+      }
+    };
+  }
```

Comment:
I've defined these listeners in the constructor so they don't get reinitialised when `create()` runs for the second or third times.